Create gnomadv4_extract_test.py

tiboloic opened the pull request:
Script to extract variants for a panel of 10 genes from gnomAD v4
and save as a local small hail Table (rows only)

Should be run on a Dataproc in USA

Feedback needed about where to save resulting small table, with options being

- local disk
- bucket in USA (to be created, to avoid egress)

tiboloic committed:
Create gnomadv4_extract_test.py

Scripts to extract variants for a panel of 10 genes from gnomAD v4
and save as a hail Table (rows only)

lgruen commented on gnomad/gnomadv4_extract_test.py:
```diff
@@ -0,0 +1,50 @@
+big_vds = hl.vds.read_vds('gs://gnomad/v4.0/raw/exomes/gnomad_v4.0.vds')
+
+# find genomics coordinates of target transcripts and filter vds
+target_intervals = hl.experimental.get_gene_intervals(transcript_ids=TRANSCRIPTS, reference_genome='GRCh38')
```
Do we want to use a specific GENCODE version here?

tiboloic replied:
nope, I prefer to use the gnomAD standard for GRCh38

lgruen replied:
Not sure what you mean by gnomAD standard, but the GENCODE version differs for each gnomAD version: https://gnomad.broadinstitute.org/help/what-version-of-gencode-was-used-to-annotate-variants

The Hail default GENCODE version is 29, which is fairly old: https://hail.is/docs/0.2/experimental/index.html#hail.experimental.get_gene_intervals

tiboloic replied:
Yes, I am aware of that and would like to stick to the default version of GENCODE for GRCh38 V29

lgruen replied:
Why though? That's neither compatible with v2, nor v3.1, nor the VEP version which we'd presumably would be running on gnomAD v4, right?

tiboloic replied:
Change to GENCODE v39 for full compatibility with Alphafold2

lgruen commented on gnomad/gnomadv4_extract_test.py:
```diff
@@ -0,0 +1,50 @@
+
+# find genomics coordinates of target transcripts and filter vds
+target_intervals = hl.experimental.get_gene_intervals(transcript_ids=TRANSCRIPTS, reference_genome='GRCh38')
+small_vds = hl.vds.filter_intervals(big_vds, target_intervals, split_reference_blocks=False, keep=True)
```
Why not `split_reference_blocks=True`?

tiboloic replied:
would be more computationally intensive with no extra benefit (saving space by segmenting reference blocks is not beneficial here, given the densification next line)

lgruen replied:
Won't densification still give you a larger interval though if you don't split here?

tiboloic replied:
No, since we keep only variants.

lgruen replied:
Where does that filtering happen? Why doesn't this also retain ref-blocks? Did you mean to only access `vds.variant_data`?

tiboloic replied:
The filtering happens during `densify` which removes all the ref-blocks

lgruen replied:
Are you sure that's the case? E.g.:

```
>>> mt = hl.read_matrix_table('gs://cpg-tob-wgs-test/mt/v7.mt')
>>> mt = mt.head(1000)
>>> mt = hl.experimental.vcf_combiner.densify(mt)
>>> mt.row_key.show()
+---------------+------------+
| locus         | alleles    |
+---------------+------------+
| locus<GRCh38> | array<str> |
+---------------+------------+
| chr1:1        | ["N"]      |
| chr1:10001    | ["T"]      |
| chr1:10003    | ["A"]      |
| chr1:10004    | ["C"]      |
| chr1:10005    | ["C"]      |
| chr1:10006    | ["C"]      |
| chr1:10007    | ["T"]      |
| chr1:10008    | ["A"]      |
| chr1:10009    | ["A"]      |
````

As far as I understand, `densify` fills in hom-refs, but doesn't remove rows.

tiboloic replied:
Yes sure, `densify`  removes the REF blocks, the rows were removed earlier by `filter_intervals`. I do not care if there is residual REF/REF rows since I am only interested in getting AN, AC, AF

lgruen commented on gnomad/gnomadv4_extract_test.py:
```diff
@@ -0,0 +1,50 @@
+mt = hl.vds.to_dense_mt(small_vds)
+
+# get alleles frequencies
+mt = mt.transmute_entries(GT=hl.experimental.lgt_to_gt(mt.LGT, mt.LA))
```
Is that still necessary after densification?

tiboloic replied:
I believe so, densify only sorts out the missing REF alleles

lgruen replied:
Ah, right. Should this be using `hl.vds.lgt_to_gt` then (instead of `experimental`)? Alternatively, you could also split multi-allelics, which might be useful downstream anyway: https://hail.is/docs/0.2/vds/hail.vds.split_multi.html

tiboloic replied:
No need to split multi_allelics at that stage, it wastes space
`hl.vds.lgt_to_gt` applies to a vds here I have a mt

lgruen commented on gnomad/gnomadv4_extract_test.py:
```diff
@@ -0,0 +1,50 @@
+
+# get alleles frequencies
+mt = mt.transmute_entries(GT=hl.experimental.lgt_to_gt(mt.LGT, mt.LA))
+mt = hl.variant_qc(mt)
```
Before running `variant_qc`, do you also want to run something like `filter_to_adj` for at least some basic variant QC?

tiboloic replied:
`filter_to_adj` needs AC, AN, AF annotations so could be run only after `variant_qc`
at this stage I don't want to exclude any variant

lgruen replied:
Not sure I follow -- `annotate_adj` only uses `GT`, `GQ`, `DP`, and `AD`?
https://broadinstitute.github.io/gnomad_methods/_modules/gnomad/utils/annotations.html#annotate_adj

tiboloic replied:
I do not want to filter anything with criteria I do not control. no `filter_to_adj` needed at that stage

lgruen commented on gnomad/gnomadv4_extract_test.py:
```diff
@@ -0,0 +1,50 @@
+
+# keep only the rows and write down
+ht = mt.rows()
+ht.write("small_ht", overwrite=True)
```
Given how tiny this is, an AU bucket should be fine? You could use `cpg_utils.hail_batch.output_path` here if you're planning to run this through the analysis-runner.

tiboloic replied:
can we talk about this in a quick call ?

lgruen commented on gnomad/gnomadv4_extract_test.py:
```diff
@@ -0,0 +1,50 @@
+
+
+# target transcripts for panel of initial 10 genes selected by the structural constraint project
+# TODO should be in an external file
```
Maybe just a `json` file, which is super easy to read (`json.load`)?

tiboloic replied:
implemented

lgruen commented on gnomad/gnomadv4_extract_test.py:
```diff
@@ -0,0 +1,50 @@
+# LT 23/08/2022
```
git already keeps author and time information on a line-by-line basis (+ history).

tiboloic replied:
an old habit, I rather stick to it if it doesn't hurt

tiboloic committed:
Moved transcript list to json file

Moved the set of transcripts into a json file
Updated code to load transcripts from file

tiboloic committed:
get ready for analysis-runner

Added main.py to encapsulate script
Added safe guard to export mt
Changed GENCODE version to v39
Used output_path() to get path right
Corrected a transcript ID

tiboloic committed:
setup for analysis runner

Added comments
Added a filtering to remove REF pseudo variants
Added setup of dataproc (parameters) in main.py

tiboloic committed:
nits on main.py before first run

tiboloic committed:
nits

replaced " by '
persisted ht to avoid double evaluation

tiboloic committed:
nits

get GNECODE v39 from CPG reference bucket
change the name of the ht to small.ht
